src/vma/util/utils.h

```diff
@@ -396,46 +397,46 @@ static inline void create_mgid_from_ipv4_mc_ip(uint8_t *mgid, uint16_t pkey, uin
  * special design for the rx loop. 
  */
 class loops_timer {
-        public:
-                loops_timer();
-                void start();
-                int  time_left_msec();
-                void set_timeout_msec(int timeout_msec) { m_timeout_msec = timeout_msec; }
-                int  get_timeout_msec() { return m_timeout_msec; }
-                inline bool is_timeout() {
-                        if (m_timeout_msec == -1)
-                                return false;
-
-                        if (m_timer_countdown > 0) {
-                                m_timer_countdown--;
-                                return false;
-                        }
-                        //init counter
-                        m_timer_countdown = m_interval_it;
-
-                        if (!ts_isset(&m_start)) {
-                                gettime(&m_start);
-                        }
-                        //update timer
-                        gettime(&m_current);
-                        ts_sub(&m_current, &m_start, &m_elapsed);
-                        vlog_printf(VLOG_FUNC_ALL, "update loops_timer (elapsed time=%ld sec %ld usec\n", ts_to_sec(&m_elapsed), ts_to_usec(&m_elapsed));
-
-
-
-                        // test for timeout 
-                        if (m_timeout_msec <= ts_to_msec(&m_elapsed)) 
-                                return true;
-
-                        return false;
-                }
-        private:
-                timespec m_start;
-                timespec m_elapsed;
-                timespec m_current;
-                int m_interval_it;
-                int m_timer_countdown;
-                int m_timeout_msec;
+public:
+	loops_timer();
+	void start();
+	int  time_left_msec();
+	void set_timeout_msec(int timeout_msec) { m_timeout_msec = timeout_msec; }
+	int  get_timeout_msec() { return m_timeout_msec; }
+	inline bool is_timeout()
+	{
+		if (m_timeout_msec == -1)
+			return false;
+
+		if (m_timer_countdown > 0) {
+			m_timer_countdown--;
+			return false;
+		}
+		//init counter
+		m_timer_countdown = m_interval_it;
+
+		if (m_start.time_since_epoch() == steady_clock::duration::zero()) {
+			m_start = steady_clock::now();
+		}
+		// update timer
+		m_current = steady_clock::now();
+		m_elapsed = duration_cast<milliseconds>(m_current - m_start);
+
+		auto elapsed_seconds = duration_cast<seconds>(m_elapsed);
+		vlog_printf(VLOG_FUNC_ALL, "update loops_timer (elapsed time=%ld sec %ld msec\n",
+			  elapsed_seconds.count(),
+			  duration_cast<milliseconds>(m_elapsed - elapsed_seconds).count());
+
+		// test for timeout
+		return m_timeout_msec <= m_elapsed.count();
+	}
```

Comment:
I would suggest to

1. move this implementation to utils.cpp file
2. remove `using namespace std::chrono;` above
3. replace `vlog_printf(VLOG_FUNC_ALL, ...` with `__log_finer()` to remove it in compile time when not needed
4. print the m_elapsed in milliseconds without converting to seconds

Reply:
1. OK
2. I prefer to keep using the chrono namespace to keep the following code:
```
  time_point<steady_clock> m_start;
  milliseconds m_elapsed;
  time_point<steady_clock> m_current;
```
instead of:
```
  std::chrono::time_point<std::chrono::steady_clock> m_start;
  std::chrono::milliseconds m_elapsed;
  std::chrono::time_point<steady_clock> m_current;
```

Reply:
I'd prefer 3 subsequent lines with the chrono namespace instead of global "using namespace" (VMA's situation with headers is not perfect and basically most of the headers are always included)

Reply:
and also remove the using namespace from delta_timer.h too for the same reason

Reply:
OK, Done!